libs/surfaces/websockets/server.cc

```diff
@@ -279,35 +312,100 @@ WebsocketsServer::write_client (Client wsi)
 	pending.pop_front ();
 
 	unsigned char out_buf[1024];
-	size_t        len = msg.serialize (out_buf + LWS_PRE, 1024 - LWS_PRE);
+	int len = msg.serialize (out_buf + LWS_PRE, 1024 - LWS_PRE);
 
 	if (len > 0) {
 #ifndef NDEBUG
 		std::cerr << "TX " << msg.state ().debug_str () << std::endl;
 #endif
-		lws_write (wsi, out_buf + LWS_PRE, len, LWS_WRITE_TEXT);
+		if (lws_write (wsi, out_buf + LWS_PRE, len, LWS_WRITE_TEXT) != len) {
+			return 1;
+		}
 	} else {
 		PBD::error << "ArdourWebsockets: cannot serialize message" << endmsg;
 	}
 
 	if (!pending.empty ()) {
 		lws_callback_on_writable (wsi);
 	}
+
+	return 0;
 }
 
-void
-WebsocketsServer::reject_http_client (Client wsi)
+int
+WebsocketsServer::send_index_hdr (Client wsi)
+{
+	char url[1024];
+
+	if (lws_hdr_copy (wsi, url, 1024, WSI_TOKEN_GET_URI) < 0) {
+		return 1;
+	}
+
+	if (strcmp (url, "/index.json") != 0) {
+		lws_return_http_status (wsi, 404, 0);
+		return 1;
+	}
+
+	unsigned char out_buf[1024],
+		*start = out_buf,
+		*p = start,
+		*end = &out_buf[sizeof(out_buf) - 1]; 
+
+#if LWS_LIBRARY_VERSION_MAJOR >= 3
+	lws_add_http_common_headers (wsi, HTTP_STATUS_OK, "application/json",
+		LWS_ILLEGAL_HTTP_CONTENT_LEN, &p, end);
+	lws_add_http_header_by_token (wsi, WSI_TOKEN_HTTP_CACHE_CONTROL,
+		reinterpret_cast<const unsigned char*> ("no-store"), 8, &p, end);
+
+	if (lws_finalize_write_http_header (wsi, start, &p, end) != 0) {
+		return 1;
+	}
+#else
+	lws_add_http_header_status (wsi, HTTP_STATUS_OK, &p, end);
+	lws_add_http_header_by_token (wsi, WSI_TOKEN_HTTP_CONTENT_TYPE,
+			reinterpret_cast<const unsigned char*> ("application/json"), 16, &p, end);
+	lws_add_http_header_by_token (wsi, WSI_TOKEN_CONNECTION,
+			reinterpret_cast<const unsigned char*> ("close"), 5, &p, end);
+	lws_add_http_header_by_token (wsi, WSI_TOKEN_HTTP_CACHE_CONTROL,
+			reinterpret_cast<const unsigned char*> ("no-store"), 8, &p, end);
+	lws_finalize_http_header (wsi, &p, end);
+
+	int len = p - start;
+
+	if (lws_write (wsi, start, len, LWS_WRITE_HTTP_HEADERS) != len) {
+		return 1;
+	}
+#endif
+
+	lws_callback_on_writable (wsi);
+
+	return 0;
+}
+
+int
+WebsocketsServer::send_index_body (Client wsi)
 {
-	const char *html_body = "<p>This URL is not meant to be accessed via HTTP; for example using"
-		" a web browser. Refer to Ardour documentation for further information.</p>";
-	lws_return_http_status (wsi, 404, html_body);
+	std::string index = _resources.scan ();
+
+	char body[MAX_INDEX_SIZE];
+	//lws_strncpy (body, index.c_str (), sizeof(body));
+	memset (body, 0, sizeof (body));
+	strncpy (body, index.c_str (), sizeof(body) - 1);
+	int len = strlen (body);
```

Comment:
Why is this needed?

Can't you just use `index.c_str(), index.length()` ?

Reply:
True, no need to copy that string. I guess I copied this from the pure websockets data write part where you need to provide a buffer with initial padding, and didn't think too much. Good catch.

Reply:
Just pushed separate commits. Was going to address this in a further new commit then realized that `lws_write()` wants a pointer to mutable data and `c_str()` returns const. What's the recommended strategy there? a C cast (in the case of HTTP writes it can be assumed that libwebsockets does not touch the buffer) or blindly adhere to the API and keep the copy?